### ISSUE TITLE
Bug: turning off all of no-unsafe-(argument|assignment|return) makes some variables incorrectly detected as any

### ISSUE THREAD

**Issue by tkalliom** [typescript-eslint/typescript-eslint#5694]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Issue Description

When all of `no-unsafe-argument`, `no-unsafe-assignment` and `no-unsafe-return` are turned off, variables are sometimes incorrectly detected as any. For the below code, `groupAssociation` is considered `any` by typescript-eslint, even though TypeScript correctly assigns the type `GroupAssociation | undefined` to it. This results in false positives from e.g. `strict-boolean-expressions`. I can not give a concise summary for what is required, and even code that comes later affects this: if either of the assignments to `groupChoice` is removed, the problem no longer occurs. A full repository is linked.

```ts
import { strict as assert } from "assert";

type GroupAssociation = { enrollmentId: number; studentId: string; groupId: number };

export const findEnrollmentInformation: () => unknown = function () {
  const courseInfoById = new Map<string, { groups: { id: number }[] }>();
  const groupAssociations: GroupAssociation[] = [];
  const enrollmentRows: { id: number; studentId: string; courseId: string }[] = [];

  return enrollmentRows.map(
    function ({ id, studentId, courseId }) {
      let groupChoice: number | undefined;
      const groupAssociation = groupAssociations.find(a => a.enrollmentId === id);
      if (groupAssociation) {
        const courseInfo = courseInfoById.get(courseId);
        assert(courseInfo !== undefined, `Expected course ${courseId}, which student ${studentId} is enrolled in, to be present`);

        const { groupId } = groupAssociation;
        const group = courseInfo.groups.find(g => g.id === groupId);
        if (group === undefined) {
          console.warn(`Enrollment ${id} has group ${groupId} for student ${studentId}, but such group does not exist under course ${courseId}`);
          groupChoice = -1;
        } else {
          groupChoice = group.id;
        }
      }
      return { studentId, courseId, groupChoice };
    }
  );
};

```

### Reproduction Repository Link

https://github.com/tkalliom/typescript-eslint-incorrect-any

### Repro Steps

1) clone the repo
2) `yarn install`
3) `yarn lint`

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.38.1` |
| `@typescript-eslint/parser`            | `5.38.1` |
| `TypeScript`                           | `4.8.4` |
| `ESLint`                               | `8.24.0` |
| `node`                                 | `16.17.1` |

**Comment by JoshuaKGoldberg** [comment]
Lovely. In theory, rules shouldn't interact like this - they should all receive the same types for nodes from the underlying TypeScript program APIs. This kind of spooky entanglement is often a bug in TypeScript itself.

Investigation required - what's going on here?

**Comment by bradzacher** [comment]
Probably related to #5364 or #4689 - maybe the issue I filed with TS https://github.com/microsoft/TypeScript/issues/48313?
It's a bug in TS somewhere yeah.

If you add an explicit return type to the `.map` callback then the problem goes away.
If you convert the same code to a loop then the problem goes away.

It's something funky in TS.
